R/git.R

```diff
@@ -101,12 +101,41 @@ git_branches_no_merged <- function(root = NULL, include_master = FALSE) {
                       root = root, check = TRUE)$output
     branches <- c(master, branches)
   }
-  branches <- read.table(text = branches, stringsAsFactors = FALSE, sep = ",",
-                         col.names = c("name", "last_commit"))
+  branches <- utils::read.table(text = branches, stringsAsFactors = FALSE,
+                                sep = ",", col.names = c("name", "last_commit"))
   branches <- branches[branches$name != "gh-pages", ]
-  branches$last_commit_age <-
-    rep(as.integer(Sys.time()), nrow(branches)) - branches$last_commit
-  times <- as.POSIXct(branches$last_commit, origin = "1970-01-01", tz = "UTC")
-  branches$last_commit <- strftime(times)
+  branches$last_commit_age <- calculate_age(branches$last_commit)
+  branches$last_commit <- convert_unix_to_iso_time(branches$last_commit)
   branches
 }
+
+## This gets last 25 commits from master
+## if not master then gets the unmerged commits (limit 25)
+git_commits <- function(branch, root = NULL) {
+  if (branch == "master") {
+    args <- c("log", "--pretty='%h,%cd'", "--date=unix", "--max-count=25",
+              sprintf("refs/remotes/origin/%s", branch))
+  } else {
+    remote_branch <- sprintf("refs/remotes/origin/%s", branch)
+    args <- c("log", "--pretty='%h,%cd'", "--date=unix", "--max-count=25",
+              sprintf("--cherry refs/remotes/origin/master...", remote_branch),
+              remote_branch)
+  }
+  commits <- git_run(args, root = root, check = TRUE)$output
+  commits <- utils::read.table(text = commits, stringsAsFactors = FALSE,
+                               sep = ",", col.names = c("id", "date_time"))
+  commits$age <- calculate_age(commits$date_time)
+  commits$date_time <- convert_unix_to_iso_time(commits$date_time)
+  ## ID can be parsed as an integer by read.table if by chance the id contains
+  ## only numbers
+  commits$id <- as.character(commits$id)
+  commits
+}
+
+calculate_age <- function(times) {
+  rep(as.integer(Sys.time()), length(times)) - times
```

Comment:
This needs a unit on it:

```
now <- Sys.time()
as.integer(now - (now - 10000)) # 2 (because close to 2 hours)
as.integer(as.numeric(now - (now - 10000), "secs")) # 10000
```

(Could you add a unit test too when fixing?)

Reply:
I think the previous impl avoids this as we cast to int before doing the calculation so I was having trouble recreating this in a test. I've updated the impl anyway as better to be explicit about the conversion and added a test.